deploy-agent/deployd/common/stats.py

```diff
@@ -28,69 +49,287 @@ def __exit__(self, stats, sample_rate=1.0, tags=None):
+    """
+
+    def __init__(self, mtype=None, name=None, value=None, sample_rate=None, tags=None, ins=None):
+        self.mtype = mtype
+        self.name = name
+        self.value = value
+        self.sample_rate = sample_rate
+        self.tags = tags
+        self.ins = ins
+        try:
+            self.JSONDecodeError = json.decoder.JSONDecodeError
+        except AttributeError:
+            # python2 support
+            self.JSONDecodeError = ValueError
+
+    def __repr__(self):
+        obj = dict()
+        obj['mtype'] = self.mtype
+        if self.value is not None:
+            # value can be 0
+            obj['value'] = self.value
+        obj['name'] = self.name
+        obj['sample_rate'] = self.sample_rate
+        obj['tags'] = self.tags
+        return obj
+
+    def serialize(self):
+        """ serialize for cache writing """
+        return json.dumps(self.__repr__())
+
+    def _deserialize(self):
+        """ read in json, setting defaults for a stat
+            return: None
+        """
+        obj = json.loads(self.ins)
+        self.mtype = obj.get('mtype', None)
+        self.name = obj.get('name', None)
+        self.value = obj.get('value', None)
+        self.sample_rate = obj.get('sample_rate', None)
+        self.tags = obj.get('tags', None)
+
+    def deserialize(self, ins=None):
+        """ attempt to deserialize
+            :param: ins json as str
+            return: bool, False on error
+        """
+        if ins:
+            self.ins = ins
+        try:
+            self._deserialize()
+            return True
+        except self.JSONDecodeError:
+            return False
+        except TypeError:
+            return False
+
+
+class MetricClient:
+    """ metrics client wrapper, enables disk cache """
+
+    def __init__(self, port=METRIC_PORT_HEALTH, cache_path=METRIC_CACHE_PATH):
+        self.port = port
+        self.cache = MetricCache(path=cache_path)
+        self.stat = None
+
+    @staticmethod
+    def _add_default_tags(tags=None):
+        """ add default tags to stats
+            :param: tags as dict
+            return: dict
+        """
+        if not __version__:
+            # defensive case, should not be hit
+            return tags
+        if __version__ and not tags:
+            tags = dict()
+        tags['deploy_agent_version'] = __version__
+        return tags
+
+    @staticmethod
+    def _parse_stat(mtype=None, name=None, value=None, sample_rate=None, tags=None):
+        """ return Stat for given kwargs """
+        return Stat(mtype=mtype,
+                    name=name,
+                    value=value,
+                    sample_rate=sample_rate,
+                    tags=tags)
+
+    def _send_mtype(self):
+        """ send metric to sc using corrected
+            calls per metric type
+        """
+        func = getattr(sc, self.stat.mtype)
+        func_v2 = getattr(sc_v2, self.stat.mtype)
+
+        if self.stat.mtype == 'increment':
```

Comment:
use enum for mtype.

Reply:
the data time enum was introduced to the standard library in python 3.4